cpp/arcticdb/codec/codec.cpp

```diff
@@ -147,44 +138,68 @@ std::optional<google::protobuf::Any> decode_metadata_from_segment(const Segment
     return decode_metadata(hdr, data, begin);
 }
 
-Buffer decode_encoded_fields(
-    const arcticdb::proto::encoding::SegmentHeader& hdr,
-    const uint8_t* data,
-    const uint8_t* begin ARCTICDB_UNUSED) {
-        ARCTICDB_TRACE(log::codec(), "Decoding encoded fields");
-        MetaBuffer meta_buffer;
-        std::optional<util::BitMagic> bv;
-        if(hdr.has_column_fields()) {
-            constexpr auto type_desc = encoded_blocks_type_desc();
-            decode_field(type_desc, hdr.column_fields(), data, meta_buffer, bv, to_encoding_version(hdr.encoding_version()));
-        }
-        ARCTICDB_TRACE(log::codec(), "Decoded encoded fields at position {}", data-begin);
-        return meta_buffer.detach_buffer();
+EncodedFieldCollection decode_encoded_fields(
+        const SegmentHeader& hdr,
+        const uint8_t* data,
+        const uint8_t* begin ARCTICDB_UNUSED) {
+    ARCTICDB_TRACE(log::codec(), "Decoding encoded fields");
+
+    util::check(hdr.has_column_fields() && hdr.column_fields().has_ndarray(), "Expected encoded field description to be set in header");
+    std::optional<util::BitMagic> bv;
+    const auto uncompressed_size = encoding_sizes::uncompressed_size(hdr.column_fields());
+    constexpr auto type_desc = encoded_fields_type_desc();
+    Column encoded_column(type_desc, uncompressed_size, false, false);
+    decode_ndarray(type_desc, hdr.column_fields().ndarray(), data, encoded_column, bv, hdr.encoding_version());
+
+    ARCTICDB_TRACE(log::codec(), "Decoded encoded fields at position {}", data-begin);
+    return {std::move(encoded_column.release_buffer()), std::move(encoded_column.release_shapes())};
+}
+
+std::shared_ptr<arcticdb::proto::descriptors::FrameMetadata> extract_frame_metadata(
+    SegmentInMemory& res) {
+    auto output = std::make_shared<arcticdb::proto::descriptors::FrameMetadata>();
+    util::check(res.has_metadata(), "Cannot extract frame metadata as it is null");
+    res.metadata()->UnpackTo(output.get());
+    return output;
+}
+
+FrameDescriptorImpl read_frame_descriptor(
+    const uint8_t*& data) {
+    auto* frame_descriptor = reinterpret_cast<const FrameDescriptorImpl*>(data);
+    data += sizeof(FrameDescriptorImpl);
+    return *frame_descriptor;
+}
+
+SegmentDescriptorImpl read_segment_descriptor(
+    const uint8_t*& data) {
+    util::check_magic<SegmentDescriptorMagic>(data);
+    auto* frame_descriptor = reinterpret_cast<const SegmentDescriptorImpl*>(data);
+    data += sizeof(SegmentDescriptorImpl);
+    return *frame_descriptor;
 }
 
-std::optional<FieldCollection> decode_index_fields(
-    const arcticdb::proto::encoding::SegmentHeader& hdr,
+std::shared_ptr<FieldCollection> decode_index_fields(
```

Comment:
Was there any particular reason to switch to `shared_ptr` I'm not sure we need heap allocation from this one.

Reply:
The index descriptors get passed around a lot, sometimes duplicated etc, and they are both a) generally immutable and b) very expensive to copy. We were jumping through a load of weird hoops to mostly move them and occasionally clone them, it's much easier to just heap allocate once and pass pointers around.